File: Valid Palindrome II.py
# https://leetcode.com/problems/valid-palindrome-ii/

class Solution:
    def validPalindrome(self, s: str) -> bool:
        
        # If "s" is 1 letter long, return true
        if len(s) == 1:
            return True
        
        # Create pointers
        frontPtr = 0
        backPtr = len(s) - 1
        
        # Traverse string
        while frontPtr < backPtr:
            
            # If both letters match, move to next iteration
            if s[frontPtr] == s[backPtr]:
                frontPtr += 1
                backPtr -= 1
                continue
            
            # If both letters don't match, then parse string
            # NOTE: Only 1 letter be dropped at most. If the substring isn't a palindrome after skipping 1 letter, then
            #       entire string isn't a palindrome.
            #       Also, string slicing is [) where the last digit is non-inclusive.
            leftStr = s[frontPtr:backPtr]
            rightStr = s[frontPtr + 1:backPtr + 1]
            
            # Reverse both substrings
            if leftStr == leftStr[::-1] or rightStr == rightStr[::-1]:
                return True
            else:
                return False
        
        # If s is a palindrome, then program will exit while-loop.
        return True

File: Valid Palindrome II.java
/*
* https://leetcode.com/problems/valid-palindrome-ii/
* 
* Time complexity: O(n + 2(n - 1)) = O(n + 2n - 2) = O(2n + n) = O(n)
* Space complexity: O(1)
*/

class Solution {
    
    public boolean isPalindrome(String s, int frontPtr, int backPtr) {
         
        // OBJECTIVE: Check if substring is a palindrome
        
        // Traverse string
        while (frontPtr < backPtr) {
            
            // If both letters match, skip to next iteration
            if (s.charAt(frontPtr) == s.charAt(backPtr)) {
                frontPtr++;
                backPtr--;
                continue;
            }
            return false;
        }
        return true;
    }
    
    public boolean validPalindrome(String s) {
        
        // If length of s is 1, then it is a palindrome
        if (s.length() == 1) {return true;}
        
        // Create 2 pointers
        int frontPtr = 0;
        int backPtr = s.length() - 1;
        
        // Traverse string
        while (frontPtr < backPtr) {
            
            // If both letters match, skip to next iteration
            if (s.charAt(frontPtr) == s.charAt(backPtr)) {
                frontPtr++;
                backPtr--;
                continue;
            }
            
            // Compare left and right substring
            boolean leftStr = isPalindrome(s, frontPtr + 1, backPtr);
            boolean rightStr = isPalindrome(s, frontPtr, backPtr - 1);
            
            return leftStr || rightStr;
        }
        return true;
    }
}
